File: capitolo_1/octave/Esempio1_18.m
function Esempio1_18()
  close all
  clear all
  clc
  % Si voglia implementare uno script che secondo l'esempio riporduca i grafici
  % della Figura 1.17a e 1.17b

  %
  % Dati 1.17: Processo massa-molla nominale sotto l'azione del controllore ad isteresi:
  %   Fm(t) = Fm2 , e>E2
  %         = Fm1 , e<E1
  % Si rappresentino:
  %   a) transitorio della posizione
  %   b) transitorio della forza motrice
  %
  Mbar = 1; %Valore scelto di progetto
  hbar = 3; %Valore scelto di progetto
  kbar = 1; %Valore scelto di progetto
  sbar = 1; %Valore scelto di progetto
  Febar = 0; %Valore scelto di progetto
  E1=-0.02; %Valore scelto di progetto
  E2=0.04; %Valore scelto di progetto
  Fm1=-1; %Valore scelto di progetto
  Fm2=2; %Valore scelto di progetto
  Fm=0; %Valore scelto di progetto
  s0 = 0; %Condizioni iniziali nulle sulla posizione
  s0dot = 0; %Condizioni iniziali nulle sulla velocità
  x0 = [s0 s0dot]; %Vettore condizioni iniziali

  dt = 0.001; %Definizione del tempo
  t = 0:dt:10;

  % Figura 1.15a
  [t,x] = ode45(@controlloreAnelloChiuso, t, [x0], Mbar, hbar, kbar, Febar, sbar, E1, Fm1, E2, Fm2, Fm);
  transitorioPosizione = x(:,1);
  transitorioVelocita = x(:,2);
  transitorioFm = zeros(size(transitorioPosizione));

  for i = 1 : length(transitorioFm)
    errore = sbar - transitorioPosizione(i);
    if errore>=E2
      transitorioFm(i)=Fm2;
    elseif errore<=E1
      transitorioFm(i)=Fm1;
    elseif transitorioVelocita(i)>=0
      transitorioFm(i)=Fm2;
    else
      transitorioFm(i)=Fm1;
    end
  endfor

  figure(1)
  plot(t,transitorioPosizione, [0 10], [1 1], 'k');
  grid;
  xlabel('Tempo (t)')
  ylabel('Stato s')
  title('Evoluzione dello stato s per C.I. nulle')

  figure(2)
  plot(t,transitorioFm, [0 10], [0 0], 'k');
  grid;
  xlabel('Tempo (t)')
  ylabel('Forza motrice')
  title('Evoluzione della Fm per C.I. nulle')

endfunction

function [xp] = controlloreAnelloChiuso(t, x, M, h, k, Fe, s, E1, Fm1, E2, Fm2, Fm)
  % Questa funzione implementa l'evoluzione di stato per l'esempio descritto.
  % Data l'equazione
  %   M * sddot(t) = -h * sdot(t) - k * s(t) + Fm(t) - Fe
  % Si considera inoltre e = sbar - s e lo stato composto da
  %   x(t) = [ s(t)  sdot(t)]
  % Pertanto l'evoluzione di stato sarà
  %   xdot(t) = [ sdot(t)
  %               ((-h/M)*sdot(t) - (k/M)*s(t) + (1/M)*Fm(t) - Fe/M ]
  % Il valore di Fm(t) è Fm2 quando l'errore supera E2 e ci rimane finchè non
  % scende sotto E1. All'inizio viene inizializzato a seconda della velocità
  e = s - x(1);
  if e>=E2
    Fm=Fm2;
  elseif e<=E1
    Fm=Fm1;
  elseif x(2)>=0
    Fm=Fm2;
  else
    Fm=Fm1;
  end

  xp = [x(2) ...
        -(h/M) * x(2) - (k/M) * x(1) + (1/M) * Fm - Fe/M ];
endfunction

File: capitolo_1/cpp/Esempio1_18.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <string>
#include <map>
#include "matplotlibcpp.h"
#include "myDiffEquation.h"

namespace plt = matplotlibcpp;

// --------------------------
// Parametri di sistema
// --------------------------
struct Params {
    double Mbar;
    double hbar;
    double kbar;
    double Fe;
    double sbar;
    double E1;
    double E2;
    double Fm1;
    double Fm2;
};

struct EditableParams {
    double Fm;
};

// --------------------------
// Equazione differenziale - Controllore ad anello chiuso
// --------------------------
std::vector<double> f_anelloChiuso(double t, const std::vector<double>& x, const Params& p, EditableParams& ep) {
    (void)t; // unused
    std::vector<double> dx(2);
    double error = p.sbar - x[0];
    if (error >= p.E2) {
        ep.Fm = p.Fm2;
    }
    else if (error <= p.E1) {
        ep.Fm = p.Fm1;
    }
    else if (x[1] >= 0) {
        ep.Fm = p.Fm2;
    }
    else {
        ep.Fm = p.Fm1;
    }

    dx[0] = x[1];
    dx[1] = -(p.hbar/p.Mbar) * x[1] -(p.kbar/p.Mbar) * x[0] + (ep.Fm/p.Mbar) - p.Fe/p.Mbar;
    return dx;
}



// --------------------------
// MAIN
// --------------------------
int main() {
    // Dati
    double Mbar = 1, hbar = 3, kbar = 1, sbar = 1, Fe = 0;
    double E1 = -0.02, E2 = 0.02, Fm1 = -1, Fm2 = 2, Fm = 0;
    double dt = 0.001;
    double t0 = 0, tf = 10.0;
    int N = static_cast<int>((tf - t0) / dt);
    std::vector<double> time(N + 1);
    for (int i = 0; i <= N; ++i) time[i] = t0 + i * dt;

    // -----------------------------------------
    // Figura 1.17 - Transitorio posizione massa-molla, C.I. nulle, sotto azione del controllore.
    // -----------------------------------------
    std::vector<double> transitorioPosizione;
    std::vector<double> transitorioFmotrice;
    Params p = Params{Mbar, hbar, kbar, Fe, sbar, E1, E2, Fm1, Fm2};
    EditableParams ep = EditableParams{Fm};

    auto traj = rungeKutta4(f_anelloChiuso, {0.0, 0.0}, t0, tf, dt, p, ep);
    // Estrae la posizione
    for (auto& x : traj) transitorioPosizione.push_back(x[0]);
    // Estrae la forza motrice
    for (int i = 0; i < traj.size(); ++i) {
        double error = sbar - traj[i][0];
        double Fm;
        if (error >= E2) {
            Fm = Fm2;
        }
        else if (error <= E1) {
            Fm = Fm1;
        }
        else if (traj[i][1] >= 0) {
            Fm = Fm2;
        }
        else {
            Fm = Fm1;
        }
        transitorioFmotrice.push_back(Fm);
    }

    // Plot Figura 1.17a - Transitorio posizione
    plt::figure(1);
    plt::plot(time, transitorioPosizione);
    plt::xlabel("Tempo (t)");
    plt::ylabel("Posizione s");
    plt::title("Transitorio della posizione - C.I. nulle");
    plt::legend();
    plt::grid(true);

    // Plot Figura 1.17b - Transitorio forza motrice
    plt::figure(2);
    plt::plot(time, transitorioFmotrice);
    plt::xlabel("Tempo (t)");
    plt::ylabel("Forza motrice Fm");
    plt::title("Transitorio della forza motrice - C.I. nulle");
    plt::legend();
    plt::grid(true);

    plt::show();

    return 0;
}
